File: Roboti/app/src/main/java/cz/muni/fi/umimecesky/roboti/activity/ListCategoriesActivity.java
package cz.muni.fi.umimecesky.roboti.activity;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ListView;
import android.widget.Toast;

import java.io.Serializable;
import java.util.List;

import cz.muni.fi.umimecesky.roboti.R;
import cz.muni.fi.umimecesky.roboti.adapterlistener.CategoryAdapter;
import cz.muni.fi.umimecesky.roboti.db.CategoryDbHelper;
import cz.muni.fi.umimecesky.roboti.pojo.Category;
import cz.muni.fi.umimecesky.roboti.utils.Utils;

import static cz.muni.fi.umimecesky.roboti.utils.Constant.LAST_FILLED_WORD;
import static cz.muni.fi.umimecesky.roboti.utils.Constant.TICKED_CATEGORIES;

public class ListCategoriesActivity extends AppCompatActivity {

    private CategoryDbHelper categoryHelper;
    private CategoryAdapter dataAdapter;

    private Button tickAllButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_list_categories);

        categoryHelper = new CategoryDbHelper(this);

        tickAllButton = (Button) findViewById(R.id.tickAll);

        displayListView();
        setButtonClick();
    }

    private void displayListView() {
        List<Category> categories = categoryHelper.getAllCategories();

        dataAdapter = new CategoryAdapter(this, this, R.layout.category_info, categories, tickAllButton);
        ListView listView = (ListView) findViewById(R.id.listView);
        listView.setAdapter(dataAdapter);
    }


    private void setButtonClick() {

        Button nextButton = (Button) findViewById(R.id.next);
        nextButton.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View v) {

                List<Category> selectedCategories = dataAdapter.getSelectedCategories();
                if (selectedCategories.isEmpty()) {
                    Toast.makeText(ListCategoriesActivity.this, "Zvolte alespoň jednu kategorii",
                            Toast.LENGTH_LONG).show();
                    return;
                }

                Intent intent = new Intent(getBaseContext(), TrainingActivity.class);
                intent.putExtra(TICKED_CATEGORIES, (Serializable) selectedCategories);

                SharedPreferences sharedPref = Utils.getSharedPreferences(getBaseContext());
                sharedPref.edit().putString(LAST_FILLED_WORD, null).apply();
                startActivity(intent);

            }
        });


        tickAllButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                boolean allChecked = dataAdapter.areAllChecked();
                Log.v("allChecked", String.valueOf(allChecked));
                dataAdapter.setAll(!allChecked);

            }
        });

        Button backButton = (Button) findViewById(R.id.back);
        backButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

    }

}

File: app/src/main/java/cz/muni/fi/umimecesky/game/practise/ListCategoriesActivity.kt
/*
 * Copyright (c) 2018 Marek Sabo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *                http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cz.muni.fi.umimecesky.game.practise

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import android.view.View
import com.jaredrummler.materialspinner.MaterialSpinner
import cz.muni.fi.umimecesky.R
import cz.muni.fi.umimecesky.db.helper.categoryOpenHelper
import cz.muni.fi.umimecesky.prefs
import cz.muni.fi.umimecesky.game.shared.util.Constant
import cz.muni.fi.umimecesky.game.shared.util.Constant.TICKED_CATEGORIES
import cz.muni.fi.umimecesky.game.shared.util.GuiUtil.hideNavigationBar
import kotlinx.android.synthetic.main.activity_list_categories.backButton
import kotlinx.android.synthetic.main.activity_list_categories.listView
import kotlinx.android.synthetic.main.activity_list_categories.nextButton
import kotlinx.android.synthetic.main.activity_list_categories.roundsSpinner
import kotlinx.android.synthetic.main.activity_list_categories.tickAll
import org.jetbrains.anko.longToast
import org.jetbrains.anko.startActivity


class ListCategoriesActivity : AppCompatActivity() {

    private var dataAdapter: CategoryAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_list_categories)


        setSpinner()
        displayListView()
        setButtonClick()
    }

    private fun setSpinner() {
        val spinnerValues = Constant.ROUND_POSSIBILITIES

        roundsSpinner.setItems(spinnerValues)
        roundsSpinner.selectedIndex = spinnerValues.indexOf(prefs.seriesAmount)
        roundsSpinner.setOnItemSelectedListener(createSpinnerListener())
    }

    private fun createSpinnerListener(): MaterialSpinner.OnItemSelectedListener<String> =
            MaterialSpinner.OnItemSelectedListener { _, _, _, item -> prefs.seriesAmount = item }

    private fun displayListView() {
        val allCategories = categoryOpenHelper.allCategories()

        dataAdapter = CategoryAdapter(this, this, R.layout.category_info, allCategories, tickAll)
        listView.adapter = dataAdapter
    }


    private fun setButtonClick() {

        nextButton.setOnClickListener(View.OnClickListener {
            val selectedCategories = dataAdapter!!.selectedCategories
            if (selectedCategories.isEmpty()) {
                longToast(R.string.choose_at_least_one_category)
                return@OnClickListener
            }

            startActivity<TrainingActivity>(TICKED_CATEGORIES to selectedCategories)
        })


        tickAll.setOnClickListener {
            val allChecked = dataAdapter!!.areAllChecked()
            Log.v("allChecked", allChecked.toString())
            dataAdapter!!.setAll(!allChecked)
        }

        backButton.setOnClickListener { finish() }

    }

    override fun onPause() {
        super.onPause()
        prefs.checkedStates = dataAdapter!!.getCheckedStates()
    }

    override fun onResume() {
        super.onResume()
        val statesArray = prefs.checkedStates
        if (statesArray.size > 1) dataAdapter!!.setCheckedStates(statesArray)
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) hideNavigationBar()
    }
}
